### PR TITLE
Authdriver 

### LINKED ISSUE #76: (POC) (8) Manage credentials according to the cloud provider
Currently, credentials are assigned in the vars of the role. This can obviously be overwritten by global vars or possibly in the linchpin_config.yml. However, many of the cloud providers (openstack, rackspace, aws/ec2, gce) have alternative ways of authenticating with configuration files they already provide.

The rationale here is that users will want to utilize things they already know. It isn't advantageous to add complexity by providing another way to configure credentials. We should provide a way to access the configs that already exist or are already documented.

As an example, check out http://docs.openstack.org/developer/os-client-config/.

### PR DESCRIPTION
**Pull request by samvarankashyap**
## Summary:
Designed according to PR #76 discussion 
Updated schema v4 
Updated playbooks

### example topology : 
```yaml
---
    topology_name: "ex_os_server"
    resource_groups:
      - 
        resource_group_name: "testgroup1"
        res_group_type: "openstack"
        res_defs:
          - 
            res_name: "sktestinstance"
            flavor: "m1.small"
            res_type: "os_server"
            image: "rhel-6.5_jeos"
            count: 3
            keypair: "ci-factory"
            networks:
              - "atomic-e2e-jenkins-test"
              - "atomic-e2e-jenkins-test2"
            fip_pool: "10.8.172.0/22"
        credentials:
          name: "atomic_jenkins"
          profile: "default"  # optional if not mentioned it defaults to default
          auth_type: "file" # optional defaults to "file" no other implementations yet
    resource_group_vars:
      - 
        resource_group_name : "testgroup1"
        test_var1: "test_var1 msg is grp1 hello "
        test_var2: "test_var2 msg is grp1 hello "
        test_var3: "test_var3 msg is grp1 hello "
```
### cli usage:
```
linchpin rise --creds /path/to/creds
```
or 
```
export LP_CREDS='/path/to/creds'
linchpin rise
```
### Issue fixes: 
fixes #76 

Note:
** should be merged after #218 **
** Needs Rebase **
** Needs testing (for gcloud) **

### CODE REVIEW

File: linchpin/linchpin.conf [hunk 1]
```diff
@@ -36,6 +36,7 @@ schemas_folder = schemas
 
 resources_folder = resources
 inventories_folder = inventories
+credentials_folder = credentials
```
Comment (herlo) [inline linchpin/linchpin.conf]: Is this necessary? I don't see where it's currently used.

**Comment by samvarankashyap** [comment]
PR is broken after rebase, will be fixing it soon .

**Comment by samvarankashyap** [comment]
linchpin up on openstack
```
$ linchpin --creds-path ./credentials/ up openstack
target: openstack, action: up
PLAY [schema check and Pre Provisioning Activities on topology_file] ***********
...
TASK [inventory_gen : Updating inventory_file with the absolute path] **********
ok: [localhost]

TASK [inventory_gen : Generate Generic Inventory] ******************************
ok: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=33   changed=3    unreachable=0    failed=0   
```
linchpin destroy on openstack 
```
$ linchpin --creds-path ./credentials/ destroy openstack
target: openstack, action: destroy


PLAY [schema check and Pre Provisioning Activities on topology_file] ***********

TASK [common : use linchpin_config if provided] ********************************
skipping: [localhost]
...
...
PLAY RECAP *********************************************************************
localhost                  : ok=27   changed=2    unreachable=0    failed=0   
```

File: linchpin/api/__init__.py [hunk 2]
```diff
@@ -188,6 +188,8 @@ def run_playbook(self, pinfile, targets='all', playbook='up'):
         if self.ctx.cfgs.get('ansible'):
             ansible_console = ast.literal_eval(self.ctx.cfgs['ansible'].get('console', 'False'))
 
+        self.ctx.evars['creds_path'] = self.ctx.creds_path
```
Comment (herlo) [inline linchpin/api/__init__.py]: Use set_evar here. Failed CI due to the value not being there. set_evar makes sure to create the structure for you if it isn't there.

File: linchpin/api/__init__.py [hunk 3]
```diff
@@ -188,7 +188,7 @@ def run_playbook(self, pinfile, targets='all', playbook='up'):
         if self.ctx.cfgs.get('ansible'):
             ansible_console = ast.literal_eval(self.ctx.cfgs['ansible'].get('console', 'False'))
 
-        self.ctx.evars['creds_path'] = self.ctx.creds_path
+        self.set_evar('creds_path', self.ctx.creds_path)
```
Comment (herlo) [inline linchpin/api/__init__.py]: This does not help. self.ctx.creds_path is being set into the same place. Why is this necessary?

**Comment by samvarankashyap** [comment]
linchpin up aws # with env vars 
```
linchpin up aws
target: aws, action: up


PLAY [schema check and Pre Provisioning Activities on topology_file] ***********

TASK [common : use linchpin_config if provided] ********************************
skipping: [localhost]
...
TASK [inventory_gen : Generate Generic Inventory] ******************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=33   changed=3    unreachable=0    failed=0   
```
linchpin destroy aws 
```
$ linchpin destroy aws
target: aws, action: destroy


PLAY [schema check and Pre Provisioning Activities on topology_file] ***********

TASK [common : use linchpin_config if provided] ********************************
skipping: [localhost]

TASK [common : set lp_path] ****************************************************
skipping: [localhost]
...
TASK [inventory_gen : Generate Generic Inventory] ******************************
skipping: [localhost]
PLAY RECAP *********************************************************************
localhost                  : ok=28   changed=2    unreachable=0    failed=0 
```

**Comment by samvarankashyap** [comment]
linchpin up aws # with --creds-path
```
$ linchpin --creds-path /home/srallaba/workspace/creds/ up aws
target: aws, action: up
PLAY [schema check and Pre Provisioning Activities on topology_file] ***********
TASK [common : use linchpin_config if provided] ********************************
skipping: [localhost]
...
TASK [inventory_gen : Generate Generic Inventory] ******************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=33   changed=4    unreachable=0    failed=0   
```
linchpin destroy # with --creds-path
```
$ linchpin --creds-path /home/srallaba/workspace/creds/ destroy aws
target: aws, action: destroy


PLAY [schema check and Pre Provisioning Activities on topology_file] ***********

TASK [common : use linchpin_config if provided] ********************************
skipping: [localhost]
...
...
TASK [inventory_gen : Generate Generic Inventory] ******************************
skipping: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=28   changed=3    unreachable=0    failed=0   

```

**Comment by samvarankashyap** [comment]
linchpin up gcloud # with creds-path
```
$ linchpin --creds-path /home/srallaba/workspace/creds/ up gcloud
target: gcloud, action: up
PLAY [schema check and Pre Provisioning Activities on topology_file] ***********
TASK [common : use linchpin_config if provided] ********************************
skipping: [localhost]
...
...

TASK [inventory_gen : Generate Generic Inventory] ******************************
changed: [localhost]
PLAY RECAP *********************************************************************
localhost                  : ok=32   changed=4    unreachable=0    failed=0   
```
linchpin destroy gcloud # with creds-path
```
linchpin --creds-path /home/srallaba/workspace/creds/ destroy gcloud
target: gcloud, action: destroy
PLAY [schema check and Pre Provisioning Activities on topology_file] ***********
TASK [common : use linchpin_config if provided] ********************************
skipping: [localhost]
...
...
TASK [inventory_gen : Generate Generic Inventory] ******************************
skipping: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=26   changed=2    unreachable=0    failed=0   
```

**Comment by samvarankashyap** [comment]
Note: All the above tests are performed on schema_v4 topologies. As credentials attribute is introduced in schema_v4 only . 
example workspace for testing including the topologies can be found in following link. 
[linchtest](https://github.com/samvarankashyap/linchtest)

**Comment by herlo** [comment]
What is the likelihood of removing schema_v3 with this update? Probably shouldn't be maintaining two schemas as that's not a viable option.

**Comment by samvarankashyap** [comment]
@herlo I feel we should be using schema_v4 only for auth_driver because it makes no sense in supporting both the schemas from 1.0 release. However , the only consequence is to update all the examples and documentation for openstack , gcloud and aws topologies.

**Comment by herlo** [comment]
@samvarankashyap since we agree on this point, please create an issue to remove schema_v3 as part of a 1.1.0 release moving forward.

**Comment by samvarankashyap** [comment]
@herlo: ref #243

**Pull request merged**